torch/csrc/autograd/input_metadata.h

```diff
@@ -90,39 +90,42 @@ struct InputMetadata {
     return at::zeros_symint(shape_as_dim_vector(), options_);
   }
 
-  void check_nestedness_same(const at::Tensor& grad) const {
+  bool is_nestedness_same(const at::Tensor& grad) const {
     bool grad_is_subclass = grad.unsafeGetTensorImpl()->is_python_dispatch();
     bool grad_is_nested = grad.is_nested();
     bool grad_is_cpp_nested = grad_is_nested && !grad_is_subclass;
-    TORCH_CHECK(
-        grad_is_cpp_nested == is_cpp_nested_tensor() &&
-            grad_is_nested == is_nested_,
-        "grad and the input wrt the gradient that is being computed for need to be "
-        "either both nested or both non-nested tensors. Also note that nested "
-        "tensors with different layouts do not compose currently.");
+    return grad_is_cpp_nested == is_cpp_nested_tensor() &&
+        grad_is_nested == is_nested_;
   }
 
   bool is_same_shape(const at::Tensor& grad) const {
-    check_nestedness_same(grad);
+    if (!is_nestedness_same(grad)) {
+      return false;
+    }
     if (is_cpp_nested_tensor()) {
       return grad._nested_tensor_size().is_same_size(shape_as_tensor());
     }
     return grad.sym_sizes().equals(shape_as_dim_vector());
   }
   bool is_expandable_to_shape(const at::Tensor& grad) const {
-    // Currently NestedTensors are not expandable. If this support is added then
-    // updates to reduce_grad will be needed
-    check_nestedness_same(grad);
-    return grad.is_nested()
+    // Currently C++ NestedTensors are not expandable (but Python NestedTensors
+    // may be). If this support is added then updates to reduce_grad will be
+    // needed
+    if (is_nestedness_same(grad) &&
+        !grad.unsafeGetTensorImpl()->is_python_dispatch()) {
+      return false;
```

Comment:
For the case when they both non nested and not python_dispatch, should it return at::is_expandable_to(shape_as_dim_vector(), grad.sym_sizes())?

Reply:
Yup, there's a typo here. This line should be: `!is_nestedness_same(grad) && !grad.unsafeGetTensorImpl()->is_python_dispatch()`.